lib/github/merge/local.ex

```diff
@@ -0,0 +1,248 @@
+        patch = patch_link.patch
+
+        {:ok, commits} = GitHub.get_pr_commits(repo_conn, patch.pr_xref)
+        {:ok, pr} = GitHub.get_pr(repo_conn, patch.pr_xref)
+
+        {token, _} = repo_conn
+        user = GitHub.get_user_by_login!(token, pr.user.login)
+
+        git.(["fetch", "origin", patch.commit])
+        {_, 0} = git.(["merge", patch.commit])
+        %{tree: tree} = local_commit_info(workdir)
+
+        # If a user doesn't have a public email address in their GH profile
+        # then get the email from the first commit to the PR
+        user_email =
+          if user.email != nil do
+            user.email
+          else
+            Enum.at(commits, 0).author_email
+          end
+
+        # Manually create a squash commit with the directory contents after the merge
+        commit_message =
+          Batcher.Message.generate_squash_commit_message(
+            pr,
+            commits,
+            user_email,
+            toml.cut_body_after
+          )
+
+        {new_head, 0} = git.(["commit-tree", tree, "-p", prev_head, "-m", commit_message])
+        String.trim(new_head)
+      end)
+
+    git.(["checkout", head])
+    Hooks.invoke_after_merge_hook!(workdir)
+    commit_hook_changes!(workdir)
+
+    persist_local_to_github!(stmp, workdir)
+    GitHub.delete_branch!(repo_conn, stmp)
+    local_commit_info(workdir)
+  end
+
+  @spec local_commit_info(String.t()) :: %{commit: String.t(), tree: String.t()}
+  defp local_commit_info(workdir) do
+    {commit, 0} = System.cmd("git", ["log", "-n1", "--pretty=%H"], cd: workdir)
+    {tree, 0} = System.cmd("git", ["log", "-n1", "--pretty=%T"], cd: workdir)
+
+    %{commit: String.trim(commit), tree: String.trim(tree)}
+  end
+
+  @spec init_project_repo!(Project.t(), String.t()) :: :ok
+  defp init_project_repo!(project, creds) do
+    System.cmd("git", [
+      "clone",
+      "https://#{creds}@github.com/#{project.name}.git",
+      "--recursive",
+      project.name
+    ])
+
+    :ok
+  end
+
+  @spec persist_local_to_github!(String.t(), String.t()) :: :ok
+  defp persist_local_to_github!(branch_name, workdir) do
+    # Credentials already exist in the URL for 'origin' here, from the clone.
+    # Each access token is valid for an hour, so there shouldn't be a need to
+    # refresh it before pushing.
+    {_, 0} =
+      System.cmd(
+        "git",
+        ["push", "origin", "--force", "HEAD:refs/heads/#{branch_name}"],
+        cd: workdir
+      )
+
+    :ok
+  end
+
+  @doc """
+  The intended semantics are that hooks can change the repo filesystem, and
+  we'll combine those changes with our merge commit. This function is careful
+  to not add a new commit to the commit history.
+  """
+  @spec commit_hook_changes!(String.t()) :: :ok
+  def commit_hook_changes!(workdir) do
+    git = fn args -> System.cmd("git", @bors_author ++ args, cd: workdir) end
+    {last_msg, 0} = git.(["log", "-n1", "--pretty=%B"])
+    git.(["add", "-A"])
```

Comment:
check that this captures deletions

Reply:
tested, captures changes, new files, and file deletions